File: cpp/ContainerContext.cpp
///
/// ContainerContext.cpp
/// Container context management for layout tracking
///

#include "ContainerContext.hpp"

namespace margelo::nitro::cssnitro {

    // Static member initialization
    std::unordered_map<std::string, LayoutBounds> ContainerContext::_layoutMap;
    std::unordered_map<std::string, ScopeHierarchy> ContainerContext::_scopeMap;

    std::optional<std::string> ContainerContext::findInScope(const std::string &containerScope,
                                                             const std::optional<std::string> &name) {
        // If no name provided, return the container scope directly
        if (!name.has_value()) {
            return containerScope;
        }

        // Check if the scope exists
        auto scopeIt = _scopeMap.find(containerScope);
        if (scopeIt == _scopeMap.end()) {
            return std::nullopt;
        }

        const ScopeHierarchy &hierarchy = scopeIt->second;

        // Check if name exists in current scope
        if (hierarchy.names.find(name.value()) != hierarchy.names.end()) {
            return containerScope;
        }

        // If not found and parent is not "root", check parent scope
        if (!hierarchy.parent.empty() && hierarchy.parent != "root") {
            return findInScope(hierarchy.parent, name);
        }

        // Not found and we've reached root
        return std::nullopt;
    }

    void ContainerContext::setScope(const std::string &containerScope,
                                    const std::string &parent,
                                    const std::unordered_set<std::string> &names) {
        _scopeMap[containerScope] = ScopeHierarchy(parent, names);
    }

    std::optional<double> ContainerContext::getX(const std::string &containerScope,
                                                 const std::optional<std::string> &name,
                                                 reactnativecss::Effect::GetProxy &get) {
        auto foundKey = findInScope(containerScope, name);
        if (!foundKey.has_value()) {
            return std::nullopt;
        }

        // Get layout bounds
        auto it = _layoutMap.find(foundKey.value());
        if (it == _layoutMap.end()) {
            return std::nullopt;
        }

        // Track the observable through the proxy
        return get(*it->second.x);
    }

    std::optional<double> ContainerContext::getY(const std::string &containerScope,
                                                 const std::optional<std::string> &name,
                                                 reactnativecss::Effect::GetProxy &get) {
        auto foundKey = findInScope(containerScope, name);
        if (!foundKey.has_value()) {
            return std::nullopt;
        }

        auto it = _layoutMap.find(foundKey.value());
        if (it == _layoutMap.end()) {
            return std::nullopt;
        }

        return get(*it->second.y);
    }

    std::optional<double> ContainerContext::getWidth(const std::string &containerScope,
                                                     const std::optional<std::string> &name,
                                                     reactnativecss::Effect::GetProxy &get) {
        auto foundKey = findInScope(containerScope, name);
        if (!foundKey.has_value()) {
            return std::nullopt;
        }

        auto it = _layoutMap.find(foundKey.value());
        if (it == _layoutMap.end()) {
            return std::nullopt;
        }

        return get(*it->second.width);
    }

    std::optional<double> ContainerContext::getHeight(const std::string &containerScope,
                                                      const std::optional<std::string> &name,
                                                      reactnativecss::Effect::GetProxy &get) {
        auto foundKey = findInScope(containerScope, name);
        if (!foundKey.has_value()) {
            return std::nullopt;
        }

        auto it = _layoutMap.find(foundKey.value());
        if (it == _layoutMap.end()) {
            return std::nullopt;
        }

        return get(*it->second.height);
    }

    void ContainerContext::setLayout(const std::string &key,
                                     double x, double y,
                                     double width, double height) {
        LayoutBounds &bounds = _layoutMap[key];
        bounds.x->set(x);
        bounds.y->set(y);
        bounds.width->set(width);
        bounds.height->set(height);
    }

} // namespace margelo::nitro::cssnitro

File: src/native/js-lib/ContainerContext.ts
/**
 * ContainerContext: Container context management for layout tracking
 * Converted from cpp/ContainerContext.cpp
 */

import type { GetProxy } from "./Effect";
import { Observable } from "./Observable";

interface LayoutBounds {
  x: Observable<number>;
  y: Observable<number>;
  width: Observable<number>;
  height: Observable<number>;
}

interface ScopeHierarchy {
  parent: string;
  names: Set<string>;
}

// Replace static-only class with a module-level singleton object (closure over private maps)
const layoutMap = new Map<string, LayoutBounds>();
const scopeMap = new Map<string, ScopeHierarchy>();

export const ContainerContext = {
  findInScope(containerScope: string, name?: string): string | undefined {
    if (!name) {
      return containerScope;
    }
    const hierarchy = scopeMap.get(containerScope);
    if (!hierarchy) {
      return undefined;
    }
    if (hierarchy.names.has(name)) {
      return containerScope;
    }
    if (hierarchy.parent && hierarchy.parent !== "root") {
      return this.findInScope(hierarchy.parent, name);
    }
    return undefined;
  },

  setScope(containerScope: string, parent: string, names: Set<string>): void {
    scopeMap.set(containerScope, { parent, names });
  },

  getX(
    containerScope: string,
    name: string | undefined,
    get: GetProxy,
  ): number | undefined {
    const foundKey = this.findInScope(containerScope, name);
    if (!foundKey) return undefined;
    const bounds = layoutMap.get(foundKey);
    if (!bounds) return undefined;
    return get(bounds.x);
  },

  getY(
    containerScope: string,
    name: string | undefined,
    get: GetProxy,
  ): number | undefined {
    const foundKey = this.findInScope(containerScope, name);
    if (!foundKey) return undefined;
    const bounds = layoutMap.get(foundKey);
    if (!bounds) return undefined;
    return get(bounds.y);
  },

  getWidth(
    containerScope: string,
    name: string | undefined,
    get: GetProxy,
  ): number | undefined {
    const foundKey = this.findInScope(containerScope, name);
    if (!foundKey) return undefined;
    const bounds = layoutMap.get(foundKey);
    if (!bounds) return undefined;
    return get(bounds.width);
  },

  getHeight(
    containerScope: string,
    name: string | undefined,
    get: GetProxy,
  ): number | undefined {
    const foundKey = this.findInScope(containerScope, name);
    if (!foundKey) return undefined;
    const bounds = layoutMap.get(foundKey);
    if (!bounds) return undefined;
    return get(bounds.height);
  },

  setLayout(
    key: string,
    x: number,
    y: number,
    width: number,
    height: number,
  ): void {
    let bounds = layoutMap.get(key);
    if (!bounds) {
      bounds = {
        x: Observable.create(x),
        y: Observable.create(y),
        width: Observable.create(width),
        height: Observable.create(height),
      };
      layoutMap.set(key, bounds);
    } else {
      bounds.x.set(x);
      bounds.y.set(y);
      bounds.width.set(width);
      bounds.height.set(height);
    }
  },
};
